Lombiq.Tests.UI/Services/UITestExecutionSession.cs

```diff
@@ -149,18 +149,16 @@ private async ValueTask ShutdownAsync()
 
         private async Task<List<BrowserLogMessage>> GetBrowserLogAsync(RemoteWebDriver driver)
         {
-            if (_browserLogMessages != null) return _browserLogMessages;
-
-            var allMessages = new List<BrowserLogMessage>();
+            _browserLogMessages ??= new List<BrowserLogMessage>();
 
             foreach (var windowHandle in _context.Driver.WindowHandles)
```

Comment:
This causes an exception for certain tests: https://ci.lombiq.com/buildConfiguration/OrchardExtensions_OSOCE_Developer_RunUiTests/922716?buildTab=overview&showRootCauses=false&expandBuildTestsSection=true&expandBuildChangesSection=true#%2FFailureDumps%2FBasicOrchardFeaturesShouldWork(browser_-Chrome)%2FAttempt%201%2FDebugInformation

Reply:
Firstly, that fix should not have been done in scope of this PR. As it turned out, browser logs assertions was almost not working. Only first chunk of log was asserted. We fixed that code, and some old tests became to fail, as they were implemented regardless of browser log errors, and that fix uncovered them.

I made a fix for `BasicOrchardFeaturesTests.BasicOrchardFeaturesShouldWork(browser: Chrome)`.

Here are the problems with rest of the tests:

- ﻿`BasicTests.ShortcutsShouldWork(browser: Chrome)` - navigates to page "Lombiq.OSOCE.JsonEditor.Sample" that is 404.
- `ErrorHandlingTests.ErrorOnLoadedPageShouldHaltTest(browser: Chrome)` - checks app logs but doesn't consider 404 error in browser log.

Please check out those tests and, if possible, please make correct modifications.

Reply:
This exception still happens, unreliably. [Here](https://ci.lombiq.com/buildConfiguration/OrchardExtensions_OSOCE_Developer_RunUiTests/926093?showRootCauses=false&expandBuildTestsSection=true&expandBuildChangesSection=true) it happens, the same app executed again [here](https://ci.lombiq.com/buildConfiguration/OrchardExtensions_OSOCE_Developer_RunUiTests/926186?showRootCauses=false) worked.

Also, the `TestContensAdminAsMonkeyRecursivelyShouldWorkWithAdminUser` should fail but if not for the `WebDriverException ` it most of the time doesn't (but sometimes does), probably more frequently not failing in headless mode.

Reply:
I checked the logs. So basically it depends on randomness. Test fails when it gets to "/Admin/Contents/ContentItems&Options.SelectedContentType=Article" page, which is really 404. But in [your successful build](https://ci.lombiq.com/buildConfiguration/OrchardExtensions_OSOCE_Developer_RunUiTests/926186?showRootCauses=false) it didn't visit that page, but visited "/Admin/Contents/ContentItems/4xa8ds820dgqkvx6eqh0c14rwz/Edit" instead. So it seems like fine for me.

Reply:
Shouldn't it be deterministic though? Since we hard-coded the seed. And the `WebDriverException` shouldn't happen at all.

Reply:
By design it should be deterministic. But unfortunately I also noticed that Gremlins can behave differently with the same seed. Sometimes it reproduces a previous run, but sometimes doesn't. I don't know what else we can do here, and should we. It's an internal logic of Gremlins which uses [Chance JS library](https://www.npmjs.com/package/chance) for randomization.

As I understand, running with a seed is needed to rerun a failed test, which is difficult to understand the reason, locally to reproduce a failure. For most failures a reason of failure should be clear by screenshots and logs. So it seems like a rare scenario for tests that have strange and unclear failure. Anyway, with few local runs you should be able to reproduce it.

Reply:
I've opened an issue about repeatability here, please chime in if necessary: https://github.com/marmelab/gremlins.js/issues/185

The `WebDriverException` issue remains though, as also visible in [the latest run](https://ci.lombiq.com/buildConfiguration/OrchardExtensions_OSOCE_Developer_RunUiTests/926713?showRootCauses=false&expandBuildChangesSection=true&expandBuildTestsSection=true).

Reply:
Please give them the information necessary for troubleshooting there.

Reply:
OK. Here are the findings on those 2 issues.

1. Gremlins randomness. I tried to reproduce our issue in laboratory conditions but failed, as that execution was deterministic. So I started to find what can be dynamic on our page under test that can make monkey testing execution differ. And I found a possible reason, which is the "Privacy consent" popup that slowly slides on a page from the right side of a screen when the page appears. ![image](https://user-images.githubusercontent.com/10154945/152176851-30b23805-f047-432e-9239-1fee0dbbd73a.png)
When I added temporarily `Thread.Sleep(5000);` after page load and before Gremlins script execution, to make the page frozen (popup stopped animating), the test became deterministic. So this popup sometimes can get clicks and sometimes not, depending on milliseconds when Gremlins script started and popup started its movement. It makes the page test non-deterministic. Usually, anyway it is recommended to disable such popups during UI test automation, as they can make tests flaky. The same for our case. So here are the possible solutions that I see:
   a) To fix our randomness issue, we can disable that popup at all on all pages.
   b) Or leave it, but we should consider that monkey testing can be a bit random when using the same seed.
   c) Or add a handler on every page load that will wait for that popup to finish its animation.

2. `WebDriverException`. This error was uncovered after we fixed browser log caching. I spent hours debugging and finding the reason of that. So basically the reason was in `UITestExecutionSession` and failed test rerun. When a test fails and another attempt is taken, a new second `UITestExecutionSession` instance is created. But when we get to `AssertLogsAsync`, it was executed in the scope of the first `UITestExecutionSession` instance with old context and already disposed `Driver` instance. It happened because of `_configuration.Events.AfterPageChange += _ => AssertLogsAsync();` event subscription was executed in first `UITestExecutionSession` instance, and was not unsubscribed when that instance was disposed. So logs assertion was executed in the scope of wrong and disposed `UITestExecutionSession` instance. It may be easier to understand when you look at a fix I made in commit [d8a2b54843addead81eb40cbe82dcb9422011ff8](https://github.com/Lombiq/UI-Testing-Toolbox/pull/109/commits/d8a2b54843addead81eb40cbe82dcb9422011ff8). Maybe there is a better way to fix that, as you wrote that code, you can implement a different solution I guess. And just to confirm that this `WebDriverException` gone away, I added temp `ShouldFailWith404` test method, because `TestContensAdminAsMonkeyRecursivelyShouldWorkWithAdminUser` didn't fail on TeamCity twice in a row because of a randomness issue. Here is the latest CI test result without `WebDriverException`: <https://ci.lombiq.com/viewLog.html?buildId=934218&buildTypeId=OrchardExtensions_OSOCE_Developer_RunUiTests#testNameId7163670106898338870>.

Reply:
Thank you for investigating!

1. I removed that the privacy consent banner slides in. Now it's just there right away so this shouldn't be an issue anymore.
2. Ah, that was a tricky one. Tweaked it a bit, otherwise looks good.